File: packages/data/src/services/linkedin/jobs/api-client.ts
/**
 * LinkedIn Jobs API Client (Guest API)
 * Fetches job postings from LinkedIn's guest API without authentication.
 * Limited to basic search parameters and a small number of results.
 */

import { load } from 'cheerio';

import { createLogger } from '@ajh/core';
import type { JobPosting } from '@ajh/shared';

import type { LinkedInHttpClient } from '../client/http-client.js';
import type { LinkedInSessionData } from '../session/store.js';

const logger = createLogger('linkedin.jobs.api');

// LinkedIn API pagination
const PAGE_SIZE = 25;

// Helper function to strip HTML tags
function stripHtml(html: string): string {
  return html.replace(/<[^>]*>/g, '').trim();
}

export interface JobsSearchParams {
  keywords: string;
  location?: string;
  start?: number;
  dateFilter?: string; // '24h', '8h', 'week', 'month'
  jobType?: string; // 'F' (Full-time), 'P' (Part-time), 'C' (Contract), 'T' (Temporary), 'I' (Internship), 'V' (Volunteer)
  workType?: string; // '1' (On-site), '2' (Remote), '3' (Hybrid)
  experienceLevel?: string; // '1' (Internship), '2' (Entry), '3' (Associate), '4' (Mid-Senior), '5' (Director), '6' (Executive)
  easyApply?: boolean;
  activelyHiring?: boolean;
  verified?: boolean;
  sortBy?: string; // 'DD' (Date Descending), 'R' (Relevance)
}

export interface LinkedInJobCard {
  urn: string;
  title: string;
  company: string;
  location: string;
  link: string;
  postedAt?: string;
}

export class LinkedInJobsApiClient {
  private client: LinkedInHttpClient;

  constructor(client: LinkedInHttpClient) {
    this.client = client;
  }

  /**
   * Search jobs using the guest API (no authentication required).
   * This is the most reliable method and doesn't require session cookies.
   */
  async searchGuest(params: JobsSearchParams, signal?: AbortSignal): Promise<JobPosting[]> {
    const {
      keywords,
      location,
      start = 0,
      dateFilter,
      jobType,
      workType,
      experienceLevel,
      easyApply,
      activelyHiring,
      verified,
      sortBy,
    } = params;

    const seen = new Set<string>();
    const out: JobPosting[] = [];
    const now = Date.now();

    // Map date filter to LinkedIn f_TPR parameter
    let f_TPR = '';
    if (dateFilter === '30m') f_TPR = 'r1800';
    else if (dateFilter === '1h') f_TPR = 'r3600';
    else if (dateFilter === '2h') f_TPR = 'r7200';
    else if (dateFilter === '4h') f_TPR = 'r14400';
    else if (dateFilter === '8h') f_TPR = 'r28800';
    else if (dateFilter === '24h') f_TPR = 'r86400';
    else if (dateFilter === 'week') f_TPR = 'r604800';
    else if (dateFilter === 'month') f_TPR = 'r2592000';

    // Build URL with all available filters
    const url =
      `https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search` +
      `?keywords=${encodeURIComponent(keywords)}` +
      (location ? `&location=${encodeURIComponent(location)}` : '') +
      `&start=${start}` +
      (jobType ? `&f_JT=${jobType}` : '') + // omit filter when not specified — don't default to full-time only
      (f_TPR ? `&f_TPR=${f_TPR}` : '') +
      (workType ? `&f_WT=${workType}` : '') +
      (experienceLevel ? `&f_E=${experienceLevel}` : '') +
      (easyApply ? `&f_EA=true` : '') +
      (activelyHiring ? `&f_AL=true` : '') +
      (verified ? `&f_VJ=true` : '') +
      (sortBy ? `&sortBy=${sortBy}` : '');

    logger.info({ url }, 'Fetching guest API');

    const html = await this.client.fetchHtml(url, signal);
    logger.info(
      { htmlLength: html.length, htmlPreview: html.substring(0, 500) },
      'Guest API response'
    );

    const $ = load(html);

    const cards = $('li').toArray();
    logger.info({ cardsCount: cards.length }, 'Found li elements');

    if (cards.length === 0) {
      logger.error(
        {
          url,
          htmlLength: html.length,
          htmlSample: html.substring(0, 2000),
          // Detect common failure signatures
          isAuthWall: html.includes('authwall') || html.includes('auth-wall'),
          isCaptcha: html.includes('captcha') || html.includes('CAPTCHA'),
          isChallenge: html.includes('challenge') || html.includes('bot'),
          isRedirect: html.includes('<meta http-equiv="refresh"'),
          bodySnippet: html
            .replace(/<[^>]+>/g, ' ')
            .replace(/\s+/g, ' ')
            .trim()
            .substring(0, 500),
        },
        'Guest API returned 0 job cards — possible block or HTML structure change'
      );
      return [];
    }

    for (const li of cards) {
      if (signal?.aborted) break;
      const $li = $(li);
      const link =
        $li.find('a.base-card__full-link, a.base-search-card__link').first().attr('href') ?? '';
      // The data-entity-urn is on the div inside the li, not on the li itself
      const entityUrn = $li.find('[data-entity-urn]').first().attr('data-entity-urn');
      const id = entityUrn?.split(':').pop();
      if (!id || seen.has(id)) continue;
      seen.add(id);

      const title = $li
        .find('.base-search-card__title, .job-card-container__title')
        .first()
        .text()
        .trim();
      const company = $li
        .find('.base-search-card__subtitle, .job-card-container__subtitle')
        .first()
        .text()
        .trim();
      const loc = $li
        .find('.job-search-card__location, .job-card-container__location')
        .first()
        .text()
        .trim();
      const dateAttr = $li.find('time').first().attr('datetime');

      // Stream job immediately without description for faster UX
      const posting: JobPosting = {
        id: `linkedin:${id}`,
        source: 'linkedin',
        externalId: id,
        url: link.split('?')[0] ?? '',
        title,
        company,
        location: loc,
        description: '', // Will be filled in background
        capturedAt: now,
        ...(dateAttr ? { postedAt: Date.parse(dateAttr) } : {}),
      };
      out.push(posting);

      // Fetch description in background (non-blocking)
      this.fetchJobDescription(id, signal)
        .then((description) => {
          posting.description = description;
        })
        .catch(() => {
          // Description fetch failed, leave empty
        });
    }

    logger.debug({ count: out.length }, 'Fetched jobs');
    return out;
  }

  /**
   * Search jobs with pagination support.
   * Automatically handles multiple pages based on the pages parameter.
   */
  async searchPaginated(
    params: JobsSearchParams,
    pages: number,
    signal?: AbortSignal,
    onProgress?: (progress: number) => void,
    onItem?: (item: JobPosting) => void
  ): Promise<JobPosting[]> {
    const authenticated = this.client.hasSession();
    const maxPages = Math.min(Math.max(pages, 1), 10);
    const allJobs: JobPosting[] = [];
    const seen = new Set<string>();

    logger.info(
      { keywords: params.keywords, pages: maxPages, authenticated },
      'Starting paginated search'
    );

    for (let page = 0; page < maxPages; page++) {
      if (signal?.aborted) break;

      const start = page * PAGE_SIZE;
      logger.debug({ page, start }, 'Fetching page');
      const jobs = await this.searchGuest({ ...params, start }, signal);

      logger.debug({ page, jobsCount: jobs.length }, 'Page fetched');

      // Filter duplicates and stream each job
      for (const job of jobs) {
        if (!seen.has(job.externalId ?? job.id)) {
          seen.add(job.externalId ?? job.id);
          allJobs.push(job);
          // Stream job to UI for immediate display
          if (onItem) onItem(job);
        }
      }

      // If no jobs returned, we've reached the end
      if (jobs.length === 0) {
        logger.debug(`No jobs on page ${page + 1}, stopping pagination`);
        break;
      }

      // Add delay between pages to be polite
      if (page < maxPages - 1) {
        await this.delay(500 + Math.random() * 500);
      }
    }

    if (onProgress) {
      onProgress(1);
    }

    logger.info({ totalJobs: allJobs.length }, 'Paginated search complete');
    return allJobs;
  }

  /**
   * Fetch job description from the detail page.
   * Works for both guest and authenticated requests.
   */
  private async fetchJobDescription(id: string, signal?: AbortSignal): Promise<string> {
    try {
      const url = `https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/${id}`;
      const html = await this.client.fetchHtml(url, signal);
      const $ = load(html);
      const htmlContent =
        $('.show-more-less-html__markup').html() ?? $('.description__text').html() ?? '';
      return stripHtml(htmlContent);
    } catch (error) {
      logger.warn({ error, id }, 'Failed to fetch job description');
      return '';
    }
  }

  /**
   * Delay helper for rate limiting.
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Update the HTTP client with new session data.
   */
  updateSession(sessionData: LinkedInSessionData): void {
    this.client.updateSession(sessionData);
  }
}

File: apps/tauri/src-tauri/src/scraping/linkedin/api_client.rs
#![allow(dead_code)]
use super::client::LinkedInHttpClient;
use super::session::LinkedInSessionData;
use crate::scraping::types::JobPosting;
use anyhow::Result;
use scraper::Html;
use std::collections::HashSet;

const PAGE_SIZE: usize = 25;

#[derive(Debug, Clone)]
pub struct JobsSearchParams {
    pub keywords: String,
    pub location: Option<String>,
    pub start: usize,
    pub date_filter: Option<String>,
    pub job_type: Option<String>,
    pub work_type: Option<String>,
    pub experience_level: Option<String>,
    pub easy_apply: Option<bool>,
    pub actively_hiring: Option<bool>,
    pub verified: Option<bool>,
    pub sort_by: Option<String>,
}

pub struct LinkedInJobsApiClient {
    client: LinkedInHttpClient,
}

impl LinkedInJobsApiClient {
    pub fn new(client: LinkedInHttpClient) -> Self {
        Self { client }
    }

    /// Search jobs using the guest API (no authentication required).
    pub async fn search_guest(
        &self,
        params: &JobsSearchParams,
        signal: Option<&tokio_util::sync::CancellationToken>,
    ) -> Result<Vec<JobPosting>> {
        eprintln!("[LinkedIn API] Searching with keywords: '{}', location: {:?}, start: {}", 
            params.keywords, params.location, params.start);
        let f_tpr = match params.date_filter.as_deref() {
            Some("30m") => "r1800",
            Some("1h") => "r3600",
            Some("2h") => "r7200",
            Some("4h") => "r14400",
            Some("8h") => "r28800",
            Some("24h") => "r86400",
            Some("week") => "r604800",
            Some("month") => "r2592000",
            _ => "",
        };

        let mut url = format!(
            "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?keywords={}&start={}",
            urlencoding::encode(&params.keywords),
            params.start
        );

        if let Some(ref location) = params.location {
            url.push_str(&format!("&location={}", urlencoding::encode(location)));
        }

        if let Some(ref job_type) = params.job_type {
            url.push_str(&format!("&f_JT={}", job_type));
        }

        if !f_tpr.is_empty() {
            url.push_str(&format!("&f_TPR={}", f_tpr));
        }

        if let Some(ref work_type) = params.work_type {
            url.push_str(&format!("&f_WT={}", work_type));
        }

        if let Some(ref experience_level) = params.experience_level {
            url.push_str(&format!("&f_E={}", experience_level));
        }

        if params.easy_apply.unwrap_or(false) {
            url.push_str("&f_EA=true");
        }

        if params.actively_hiring.unwrap_or(false) {
            url.push_str("&f_AL=true");
        }

        if params.verified.unwrap_or(false) {
            url.push_str("&f_VJ=true");
        }

        if let Some(ref sort_by) = params.sort_by {
            url.push_str(&format!("&sortBy={}", sort_by));
        }

        eprintln!("[LinkedIn API] Request URL: {}", url);
        
        let html = self.client.get_html(&url, signal).await?;
        
        eprintln!("[LinkedIn API] Parsing HTML response...");
        let document = Html::parse_document(&html);
        let selector = scraper::Selector::parse("li").unwrap();
        let link_selector = scraper::Selector::parse("a.base-card__full-link, a.base-search-card__link").unwrap();
        let urn_selector = scraper::Selector::parse("[data-entity-urn]").unwrap();
        let title_selector = scraper::Selector::parse(".base-search-card__title, .job-card-container__title").unwrap();
        let company_selector = scraper::Selector::parse(".base-search-card__subtitle, .job-card-container__subtitle").unwrap();
        let location_selector = scraper::Selector::parse(".job-search-card__location, .job-card-container__location").unwrap();
        let time_selector = scraper::Selector::parse("time").unwrap();

        let mut seen = HashSet::new();
        let mut jobs = Vec::new();
        let _now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as i64;

        for element in document.select(&selector) {
            if let Some(signal) = signal {
                if signal.is_cancelled() {
                    break;
                }
            }

            let link = element
                .select(&link_selector)
                .next()
                .and_then(|el| el.value().attr("href"))
                .unwrap_or("");

            let entity_urn = element
                .select(&urn_selector)
                .next()
                .and_then(|el| el.value().attr("data-entity-urn"));

            let id = entity_urn.and_then(|urn| urn.split(':').last());

            let id = match id {
                Some(id_str) => {
                    if seen.contains(id_str) {
                        continue;
                    }
                    seen.insert(id_str.to_string());
                    id_str
                }
                None => continue,
            };

            let title = element
                .select(&title_selector)
                .next()
                .map(|el| el.text().collect::<String>())
                .unwrap_or_default()
                .trim()
                .to_string();

            let company = element
                .select(&company_selector)
                .next()
                .map(|el| el.text().collect::<String>())
                .unwrap_or_default()
                .trim()
                .to_string();

            let location = element
                .select(&location_selector)
                .next()
                .map(|el| el.text().collect::<String>())
                .unwrap_or_default()
                .trim()
                .to_string();

            let date_attr = element
                .select(&time_selector)
                .next()
                .and_then(|el| el.value().attr("datetime"));

            let posted_at = date_attr.and_then(|d| chrono::DateTime::parse_from_rfc3339(d).ok());

            let job = JobPosting {
                id: format!("linkedin:{}", id),
                source: "linkedin".to_string(),
                external_id: Some(id.to_string()),
                url: link.split('?').next().unwrap_or("").to_string(),
                title,
                company,
                location: Some(location),
                description: Some(String::new()), // Will be filled in background
                captured_at: std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_millis() as i64,
                posted_at: posted_at.map(|dt| dt.timestamp_millis()),
                requirements: None,
                extra: std::collections::HashMap::new(),
            };

            jobs.push(job);
        }

        eprintln!("[LinkedIn API] Found {} jobs on this page", jobs.len());
        Ok(jobs)
    }

    /// Search jobs with pagination support.
    pub async fn search_paginated(
        &self,
        params: &JobsSearchParams,
        pages: usize,
        signal: Option<&tokio_util::sync::CancellationToken>,
        on_progress: Option<Box<dyn Fn(f32) + Send>>,
        on_item: Option<Box<dyn Fn(JobPosting) + Send>>,
    ) -> Result<Vec<JobPosting>> {
        let max_pages = pages.min(10).max(1);
        let mut all_jobs = Vec::new();
        let mut seen = HashSet::new();

        for page in 0..max_pages {
            if let Some(signal) = signal {
                if signal.is_cancelled() {
                    break;
                }
            }

            let start = page * PAGE_SIZE;
            let mut search_params = params.clone();
            search_params.start = start;

            let jobs = self.search_guest(&search_params, signal).await?;

            for job in &jobs {
                let job_id = job.external_id.clone().unwrap_or_else(|| job.id.clone());
                if !seen.contains(&job_id) {
                    seen.insert(job_id);
                    if let Some(ref on_item) = on_item {
                        on_item(job.clone());
                    }
                    all_jobs.push(job.clone());
                }
            }

            if jobs.is_empty() {
                break;
            }

            // Add delay between pages
            if page < max_pages - 1 {
                tokio::time::sleep(tokio::time::Duration::from_millis(500 + (rand::random::<u64>() % 500)))
                    .await;
            }
        }

        if let Some(on_progress) = on_progress {
            on_progress(1.0);
        }

        Ok(all_jobs)
    }

    pub fn update_session(&mut self, session_data: LinkedInSessionData) {
        self.client.update_session(session_data);
    }
}
